### PR TITLE
Type signatures

### PR DESCRIPTION
**Pull request by haskellcamargo**
This pull-request does:

- Implement parser for type signatures:
  - Literal and generic types
  - List, map and tuple types
  - Type union and intersection
  - Object types

Based on:

```pascal
type
    = literal
    | instance
    | function
    | list
    | map
    | tuple
    | object
    | generic
    | union
    | intersection
    ;

instance
    = "%" identifier
    ;

function
    = "&[" [ type { "," type } ] "]" " -> " type
    ;

list
    = "{" type "}"
    ;

map
    = "#{" type ":" type "}"
    ;

tuple
    = "#(" [ type { "," type } ] ")"
    ;

object-entry
    = identifier ":" type
    ;

object
    = "%{" [ object-entry { "," object-entry } ] "}""
    ;

literal
    = "string"
    | "number"
    | "boolean"
    | "regex"
    | atom
    ;

generic
    = identifier
    ;

union
    = type "|" type
    ;

intersection
    = type "&" type
    ;

atom
    = "@" identifier
    ;

(* implicit identifier *)
```
- Create a trait for `Parselet`, making it generic and using for types
- Refactor and isolate parsers, allowing them to be cojoined (`StmtParser`, `ExprParser`, `DeclParser`, `TypeParser` and `NameParser`)
- Create `Attachable` trait for cojoined parsers
- Implement optional signatures for `let`, `const` and functions
- Kill `open` for modules

### CODE REVIEW

File: src/ast/stmt/OpenStmt.php [hunk 1]
```diff
@@ -1,90 +0,0 @@
-<?php
```
Comment (luizperes) [inline src/ast/stmt/OpenStmt.php:1]: Why did you delete completely the `OpenStmt`? What is your plan ?

File: src/lexer/SymbolDecypher.php [hunk 2]
```diff
@@ -33,7 +33,7 @@ public static function __callStatic($method, $args)
             case '>':
                 return static::tryMatch($context, ['>>', '>=', '>>']);
             case ':':
-                return static::tryMatch($context, [':-']);
+                return static::tryMatch($context, ['::', ':-']);
```
Comment (luizperes) [inline src/lexer/SymbolDecypher.php:36]: that `::` looks so cool!

File: src/parselets/types/LiteralTypeParselet.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            'string'  => NativeQuackType::T_STR,
+            'number'  => NativeQuackType::T_NUMBER,
+            'boolean' => NativeQuackType::T_BOOL,
+            'regex'   => NativeQuackType::T_REGEX,
+            'block'   => NativeQuackType::T_BLOCK,
+            'unit'    => NativeQuackType::T_UNIT
+        ];
+        $name = $grammar->reader->resolveScope($token->getPointer());
+
+        return array_key_exists($name, $names)
```
Comment (luizperes) [inline src/parselets/types/LiteralTypeParselet.php:45]: That will have to be well documented later. It would also help to have strong error messages regarding that in order to avoid problems with generic types

File: src/parser/Attachable.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with Quack.  If not, see <http://www.gnu.org/licenses/>.
+ */
+namespace QuackCompiler\Parser;
+
+trait Attachable
+{
+    public function attachParsers($parsers)
+    {
+        foreach ($parsers as $name => $parser)
```
Comment (luizperes) [inline src/parser/Attachable.php:28]: <3

File: src/parser/DeclParser.php [hunk 5]
```diff
@@ -102,104 +106,72 @@ public function _fnStmt($is_method = false)
...
-                    $subprops[] = $this->identifier();
-                } while ($this->parser->is(';'));
-            }
-
-            $this->parser->match('}');
-        }
-
-        return new OpenStmt($name, $alias, $type, $subprops);
+        $this->reader->match(Tag::T_MODULE);
+        return new ModuleStmt($this->name_parser->_qualifiedName());
```
Comment (luizperes) [inline src/parser/DeclParser.php:158]: I see that you kept the module, How is it going to work without the `OpenStmt`?

File: src/parser/TokenReader.php [hunk 6]
```diff
@@ -27,13 +27,40 @@
...
 {
+    private $main;
     public $ast = [];
     public $grammar;
 
     public function __construct(Tokenizer $input)
     {
         parent::__construct($input);
-        $this->grammar = new Grammar($this);
+        $name_parser = new NameParser($this);
```
Comment (luizperes) [inline src/parser/TokenReader.php:37]: sooooooo good!

**Review by luizperes — APPROVED** [review]
Overall that is a beautiful PR!!! I see you used most of labels we have! :dancer:

File: src/ast/stmt/OpenStmt.php (hunk 1 above)
Comment (haskellcamargo) [reply src/ast/stmt/OpenStmt.php:1]: Rewrite it all. Maybe without `open`, but with `import`, like ES6 does.

File: src/parselets/types/LiteralTypeParselet.php (hunk 3 above)
Comment (haskellcamargo) [reply src/parselets/types/LiteralTypeParselet.php:45]: I agree. I believe an explicit declaration of generic types with `forall` may be better.

File: src/parser/DeclParser.php (hunk 5 above)
Comment (haskellcamargo) [reply src/parser/DeclParser.php:158]: `module` only defines the qualified name to identify the source, but I'm unsure if it will be kept here...
